### PR TITLE
Ignore non-standard syntax in function-url-quote

### PR DESCRIPTION
**Pull request by evilebottnawi**
https://github.com/stylelint/stylelint/issues/1133
Maybe `hasVariable` move to utils. Also we can better process all inside `url` function throw `postcss-value-parser`(ignore all word contain variable), I first tried to write exactly, but `postcss-value-parser` handling very bad word contain url(`http://some-domain.com`, return some word - ["http:","/", "/", "some-domain"] and etc). I do not know whether it is right, maybe not. If we change logic in `postcss-value-parser` we can change rule for ignore all variable and rule will work containing variable inside.

### CODE REVIEW

**Comment by jeddy3** [comment]
Thanks for looking into this.

Can you create a new util function called: `isStandardUrl` please? Similar to [`isStandardSelector`](https://github.com/stylelint/stylelint/blob/master/src/utils/isStandardSelector.js) and its [tests](https://github.com/stylelint/stylelint/blob/master/src/utils/__tests__/isStandardSelector-test.js).

Rather than *checking for* and add fixes for non-standard syntax in rules, we're going to be using the `isStandard*` utils to *whitelist* standard CSS syntax constructs e.g. *rules*, *declarations*, *selectors*, *values* **and now** *urls functions*. 

As an aside, you'll need to be account for valid urls like `url(/images/my_image@2x.png)` in your tests.

**Comment by evilebottnawi** [comment]
@davidtheclark oh, then it's hard to imagine a solution

**Comment by davidtheclark** [comment]
Am I missing something? @evilebottnawi are you commenting on @jeddy3's last sentence, "... you'll need to account for ..."?

**Comment by evilebottnawi** [comment]
@davidtheclark sorry, to @jeddy3

**Comment by davidtheclark** [comment]
I agree that it's a pretty tough problem. Having seen all this stuff makes me kind of opposed to preprocessors in a way that I would not have imagined being a few months ago. By creating distinct nonstandard languages they introduce such serious challenges to tooling. It's like these special languages have trumped effective tooling in the CSS world, and now I think with PostCSS people are realizing (at least me) that better tooling may have been a better choice all along.

**Comment by evilebottnawi** [comment]
@jeddy3 it might be worth a try `url.parse` and check on valid url before check string, what do you think?

**Comment by jeddy3** [comment]
> It's like these special languages have trumped effective tooling in the CSS world, and now I think with PostCSS people are realizing (at least me) that better tooling may have been a better choice all along.

@davidtheclark I agree 100%. I originally got involved in this project as I saw an opportunity to create an awesome tool for standard (current and future) CSS *unhindered by the need to support special and complex syntax*.  I've never used SCSS and Less, opting for vanilla CSS in the beginning and then future CSS via rework and then PostCSS. The tooling (rework and PostCSS plugins etc) always seemed so much simpler.

I know we [discussed this](https://github.com/stylelint/stylelint/issues/550#issuecomment-163060940) a few months ago. And I understood your point then that, after we'd worked so hard on the project for a year, it'd be nice if it reached as many users as possible. At the time you said:

> To that end, I would like non-standard syntax users to be able to run their (non-processed) code through stylelint whenever it's reasonable. If a single extra conditional, or something else simple, allows SCSS users to take advantage of some rule, I am willing to make that concession. If the concession turns out to be significant, though, I would err on the side of simplicity and expect SCSS users not to use the rule, or to use the ignore comments.

It feels like implementing and maintaining these concessions take ups the bulk of our time though, and it's hard to see the light at the end of the tunnel.

I see four choices going forward:

1. We pull non-standard syntax support and focus on creating a tool for the future of CSS rather than get bogged down in supporting the current, fractured, landscape of preprocessors.  It feels like monolithic  preprocessors have done their job. They innovate interesting ideas. The best of which have made it into the specs. I can see see parallels in the JS world where the best ideas from coffeescript made it in es2015.
2. We freeze non-standard syntax support where it is. Add a notice to the README that reads: "Although stylelint can read non-standard syntax its core rules are not designed to work with them. As such, if you find a particular rule is tripping up your Less or SCSS code then turn off the rule or use ignore comments.". We create a "stock response" comment along the lines of "Not all rules support non-standard syntax. Please turn off the rule". This will allow us to close down issues quickly.
3. We add the notice to the README. Field issues, but "err on the side of simplicity" and only fix the simple things. Although, historical the investigation is what take the most time (e.g. trying to understand non-outputting less mixins).
4. We carry on as we are i.e. investing a lot of time fixing all of the non-standard syntax issues. 

I kind of in favour of 1. I can get behind the idea of creating *a lean* and awesome tool for the future (which the current PostCSS people can use now). I understand this probably the most controversial of choices :)

**Comment by evilebottnawi** [comment]
@jeddy3 i agree with 1. I think that it is possible of prime task is to support the standard syntax, and only then is no standard, prioritization of such issues(bugs) would be useful. The more popular a product becomes, the more and more problems appear. It would be great to find more contributors.

**Comment by Nirazul** [comment]
(stumbled upon this by chance, sorry to add my thoughts... feel free to ignore it, as it is really your library and your choice :) )

I'd be really sorry to see the removal of non-standard "support" as your linter is far more advanced as any other (node based) solution I've tried so far. That's the reason why I've experimented with it in the past few weeks and have seen the linting errors in our codebase decrease with every version.

To be honest, most of the problems that are issues in combination with non-standard syntax seem to be in the domain of an AST parser (like postcss' [postcss-scss](https://github.com/postcss/postcss-scss) or [gonzales-pe](https://github.com/tonyganch/gonzales-pe)) that helps you differenciate between `.a-class` (selector) and `.#{$a-variable}` (selector, containing: { interpolation > variable }).
As current AST parsers lag behind possibilities of modern preprocessors, there's no solid foundation, like eslint has.

Your approach of ignoring non-standard syntax seemed perfect for the moment, as certainly more and more features of less and scss are being ported to PostCSS, like it happened from coffeescript to es6. When AST parsers start to keep up with current developments, I'm sure it's gonna get easier to maintain special syntax. Until then, well maybe it's best to deprecate (official) support for special syntax.

Anyway, thank you for the great work so far. I'm willing to contribute more if I see an opportunity for it :)

**Comment by davidtheclark** [comment]
I think it would be a bit rash to consider abandoning support for non-standard syntax right now. We have made a lot of progress, I think, and although some annoying bugs pop up now and again (and some overreactions from users about those), we don't honestly get that many bugs. Relative to the size and ambition of the project, I think we get exceptionally few.

Because we have a continually growing backlog of potential new features to build, those preprocessor bugs are annoying and seem to block forward momentum. But honestly we don't need as much forward momentum as we've had! Instead of not fixing bugs for non-standard syntax, what if we held off on adding new features, and focused for a little while on squashing those irregularities?

To rephrase & summarize that: I think that the reason preprocessor-specific bugs seem overwhelming right now is because *we are already overwhelming ourselves with the continual addition of new features*. I think they would seem a lot less annoying if we weren't pounding out new stuff all the time. Backing off on the new stuff seems to me a better pressure-relief right now than giving up on the processor stuff.

**Comment by davidtheclark** [comment]
Sorry if I seem to be contradicting myself: above I was commenting on a feeling of growing opposition to preprocessors generally and the way they've been focused on at the expense of effective CSS tooling. It's an abstract concern, not a specific concern about our project and the preprocessor stuff we deal with.

**Comment by evilebottnawi** [comment]
@davidtheclark I think we stop to adding support new preprocessors, less and sass coverange mostly projects, in generally we have no as much issue with preprocessors, example eslint have disscussion about add rules in v3 for coffeescript and typescript, I think we do all rigth

**Comment by davidtheclark** [comment]
@Nirazul Thanks for input! No need to apologize: it's all open-source so that we can hear from everybody who wants to take the time to share an opinion. I'd say the most helpful thing you can do it just help us tackle these preprocessor-specific bugs as they arise, whenever you have the time and a little extra energy.

File: src/rules/function-url-quotes/index.js [hunk 1]
```diff
@@ -12,6 +12,10 @@ export const messages = ruleMessages(ruleName, {
   expected: (q, f="url") => `Expected ${q} around ${f} argument`,
 })
 
+function hasVariable(value) {
+  return value.indexOf("$") !== -1 || value.indexOf("@") !== -1
```
Comment (davidtheclark) [inline src/rules/function-url-quotes/index.js]: If you do `return /\B@/.test(value) || /\B$/.test(value)` does that solve the problem proposed by `url(/images/my_image@2x.png)`?

Comment (evilebottnawi) [inline src/rules/function-url-quotes/index.js]: @davidtheclark I think I have a better solution today I'll PR

**Comment by evilebottnawi** [comment]
@jeddy3 @davidtheclark ping

**Comment by ntwb** [comment]
Interesting discussion, and from someone who has had little time to contribute these past couple of weeks the pace at which this repo goes forward is breathtaking. This can be a good thing, or it can be a bad thing depending on how you look at it. People sometime need time to relax ☺️, explore other projects :octocat:, spend time with family 👪 , go away for the weekend ⛺ , have a drink 🍺etc 

From my WordPress background, backward compatibility is a pretty big thing, it also involves a great deal of technical debt, WordPress has be it good or bad backward compatibility going back  5,7, 9, 10, 12 years even. If breaking backward compatibility can be avoided then it is avoided, for better or worse, but that's a discussion for elsewhere, not here.

To that avail, I'd hate to see SCSS and Less syntaxes pulled from stylelint now, they are in, they are working reasonably well IMHO and I've been working with a SCSS project this past couple of weeks and have hit upon very few issues, if I had I would have created new stylelint issues here about those issues.

So, what I was thinking might best be explained using _milestones_, how about stylelint starts to use _milestones_ and use _multiple milestones_ at that,  `CSS`, `SCSS`, `Less` milestones. This way new features, code refactoring, performance testing for CSS could all happen in a `CSS` milestone. Once the current `CSS` milestone is completed then work on the `SCSS` milestone, then the `Less` milestone, and then start again back on the next/future `CSS` milestone. If there are SCSS or Less issues whilst working on the CSS milestone then the workaround for those issues is to disable the applicable rule/s until the issues can be addressed in the next _non-standard syntax_ milestone.

**Comment by davidtheclark** [comment]
Good to hear form you @ntwb.

Speaking of spending time with family, just a heads up to all that I will be unable to keep up very much over the next week, mostly because of family visits. When I do have time to look into the repo I will be focusing on clarifying bug reports and checking PRs that fix them.

**Comment by davidtheclark** [comment]
@evilebottnawi There could also be `$` symbols in a URL, right? So should that be addressed the same way?

**Comment by evilebottnawi** [comment]
@davidtheclark i think check `syntax` inside `isStandardUrl` be good solution, because in scss all `$` be interpreted as variable

**Comment by davidtheclark** [comment]
But postcss-simple-vars uses `$`, too (that's what I use right now).

**Comment by evilebottnawi** [comment]
@davidtheclark what do you think how we can resolve this problem?

**Comment by davidtheclark** [comment]
I will try to take a closer look at the code sometime in the next few days.

**Comment by jeddy3** [comment]
> Speaking of spending time with family, just a heads up to all that I will be unable to keep up very much over the next week, mostly because of family visits.

Have a nice time :) I'm also away this Friday to Sunday, but I'll try to catch up next week.

**Comment by ntwb** [comment]
Enjoy your break everyone who is having one, I enjoyed mine 😄

**Comment by jeddy3** [comment]
> I think it would be a bit rash to consider abandoning support for non-standard syntax right now.

Every few months I throw the idea out there to test the water, but I think you're right. And you're right that the non-standard syntax bugs are relatively few :)

> But honestly we don't need as much forward momentum as we've had!

Agreed. FYI I'm around this week, but I'm away all next week at a wedding.

**Comment by davidtheclark** [comment]
I'd be ok with saying this: If the url starts with `$` or `@` or contains `+\s*[\$@]` we just ignore it. Do you think that would work?

**Comment by davidtheclark** [comment]
@evilebottnawi ping — still thinking about this one?

**Comment by evilebottnawi** [comment]
@davidtheclark yes, tomorrow i try the above solution

**Comment by jeddy3** [comment]
Lets try to get this in the next release as a lingering bug.

**Comment by davidtheclark** [comment]
Cleaned up and merged via 832d3b7aa2df29e92eea9e08d03e341e51537167

**Pull request closed without merging**